language_service/src/tests.rs

```diff
@@ -267,427 +79,268 @@ async fn target_profile_update_fixes_error() {
                             ),
                         ),
                     ],
                 ),
             ]
         "#]],
-    );
-
-    ls.close_notebook_document("notebook.ipynb", ["cell1", "cell2"].into_iter());
-
-    expect_errors(
-        &errors,
         &expect![[r#"
-            [
-                (
-                    "cell2",
-                    None,
-                    [],
-                ),
-            ]
+            SourceMap {
+                sources: [
+                    Source {
+                        name: "other_file.qs",
+                        contents: "namespace OtherFile { operation Other() : Unit {} }",
+                        offset: 0,
+                    },
+                    Source {
+                        name: "this_file.qs",
+                        contents: "namespace Foo { }",
+                        offset: 52,
+                    },
+                ],
+                entry: None,
+            }
         "#]],
     );
 }
 
-type ErrorInfo = (String, Option<u32>, Vec<compile::ErrorKind>);
-
-fn new_language_service(received: &RefCell<Vec<ErrorInfo>>) -> LanguageService<'_> {
-    let diagnostic_receiver = move |update: DiagnosticUpdate| {
-        let mut v = received.borrow_mut();
-
-        v.push((
-            update.uri.to_string(),
-            update.version,
-            update.errors.iter().map(|e| e.error().clone()).collect(),
-        ));
-    };
-
-    LanguageService::new(
-        diagnostic_receiver,
-        // these tests do not test project mode
-        // so we provide these stubbed-out functions
-        |_| Box::pin(std::future::ready((Arc::from(""), Arc::from("")))),
-        |_| Box::pin(std::future::ready(vec![])),
-        |_| Box::pin(std::future::ready(None)),
-    )
-}
-
 // the below tests test the asynchronous behavior of the language service.
 // we use `get_completions` as a rough analog for all document operations, as
 // they all go through the same `document_op` infrastructure.
 #[tokio::test]
 async fn completions_requested_before_document_load() {
     let errors = RefCell::new(Vec::new());
-    let mut ls = new_language_service(&errors);
+    let mut ls = LanguageService::default();
+    let mut worker = create_update_worker(&mut ls, &errors);
 
-    // we intentionally don't await this to test how LSP features function when
-    // a document hasn't fully loaded
-    #[allow(clippy::let_underscore_future)]
-    let _ = ls.update_document(
+    ls.update_document(
         "foo.qs",
         1,
         "namespace Foo { open Microsoft.Quantum.Diagnostics; @EntryPoint() operation Main() : Unit { DumpMachine() } }",
     );
 
+    // we intentionally don't await work to test how LSP features function when
+    // a document hasn't fully loaded
+
     // this should be empty, because the doc hasn't loaded
     assert!(ls.get_completions("foo.qs", 76).items.is_empty());
+
+    worker.apply_pending().await;
```

Comment:
Why are we awaiting after all asserts?

Reply:
Yeah, I guess I had it here for "completeness" since in real usage, updates would always be applied. And clippy didn't like that `worker` was unused. Basically this and the test below call `get_completions()` and `apply_pending()` in flipped order and I wanted that to be clear. But perhaps it looks confusing instead - I'll remove it.